File: backend/src/main/scala/domains/blog/objects/response/BlogSummary.scala
package domains.blog.objects.response

import domains.blog.objects.*

import domains.user.objects.UserIdentity
import io.circe.{Decoder, Encoder}
import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}
import shared.objects.access.ResourceVisibilityPolicy

import java.time.Instant
import scala.util.Try

/** 博客列表摘要响应，包含正文预览来源、关联题目和当前用户投票。 */
final case class BlogSummary(
  id: BlogId,
  title: BlogTitle,
  content: BlogContent,
  author: UserIdentity,
  visibilityPolicy: ResourceVisibilityPolicy,
  relatedProblems: List[BlogProblemReference],
  score: Int,
  viewerVote: Option[BlogVote],
  createdAt: Instant,
  updatedAt: Instant
)

/** 提供博客摘要 JSON codec，并显式处理 Instant 字符串格式。 */
object BlogSummary:
  private given Encoder[Instant] = Encoder.encodeString.contramap(_.toString)
  private given Decoder[Instant] = Decoder.decodeString.emap { value =>
    Try(Instant.parse(value)).toEither.left.map(_.getMessage)
  }

  given Encoder[BlogSummary] = deriveEncoder[BlogSummary]
  given Decoder[BlogSummary] = deriveDecoder[BlogSummary]

File: frontend/src/objects/blog/response/BlogSummary.ts
import type { UserIdentity } from '@/objects/user/UserIdentity'
import type { BlogContent } from '@/objects/blog/BlogContent'
import type { BlogId } from '@/objects/blog/BlogId'
import type { BlogProblemReference } from '@/objects/blog/BlogProblemReference'
import type { BlogTitle } from '@/objects/blog/BlogTitle'
import type { BlogVote } from '@/objects/blog/BlogVote'
import type { ResourceVisibilityPolicy } from '@/objects/shared/access/ResourceVisibilityPolicy'

/** 博客摘要响应；用于列表展示，包含正文摘要字段和投票状态。 */
export type BlogSummary = {
  id: BlogId
  title: BlogTitle
  content: BlogContent
  author: UserIdentity
  visibilityPolicy: ResourceVisibilityPolicy
  relatedProblems: BlogProblemReference[]
  score: number
  viewerVote: BlogVote | null
  createdAt: string
  updatedAt: string
}
